src/osd/ECBackend.cc

```diff
@@ -1605,8 +1629,10 @@ struct CallClientContexts :
   void finish(pair<RecoveryMessages *, ECBackend::read_result_t &> &in) {
     ECBackend::read_result_t &res = in.second;
     assert(res.returned.size() == to_read.size());
-    assert(res.r == 0);
-    assert(res.errors.empty());
+    if (!ec->subread_all) {
+      assert(res.r == 0);
+      assert(res.errors.empty());
+    }
```

Comment:
Accepted. (Merge with David's comment to assert out if there is less than K shards to decode).